lib/pbench/test/unit/server/query_apis/test_index_mappings.py

```diff
@@ -0,0 +1,122 @@
+import pytest
+from http import HTTPStatus
+from sqlalchemy.exc import DatabaseError
+
+from pbench.server.database.models.template import Template
+
+
+@pytest.fixture()
+def database_error(monkeypatch):
+    def raise_db_error(name: str):
+        raise DatabaseError("DB Error")
+
+    with monkeypatch.context() as m:
+        m.setattr(Template, "find", raise_db_error)
+        yield
+
+
+class TestIndexMappings:
+    """
+    Unit testing for resources/IndexMappings class.
+
+    In a web service context, we access class functions mostly via the
+    Flask test client rather than trying to directly invoke the class
+    constructor and `post` service.
+    """
+
+    def test_run_template_query(self, client, server_config, find_template):
+        """
+        Check the construction of index mappings API and filtering of the
+        response body.
+        """
+        with client:
+            response = client.get(f"{server_config.rest_uri}/index/mappings/run")
+            assert response.status_code == HTTPStatus.OK
+            res_json = response.json
+            assert res_json == {
+                "@metadata": [
+                    "controller_dir",
+                    "file-date",
+                    "file-name",
+                    "file-size",
```

Comment:
you mean returning RHS from the `find_template()`?

Reply:
We have two values -- the one which the mock feeds to the CUT and the one which the CUT returns to the test -- and the test checks that the second is a specific transformation of the first.  Since the two values are intimately related, it seems like a bad idea to have them defined in separate modules far from each other.

Although I hesitate to suggest it, my instinct is that the second one should be dynamically generated by the test from the first one.  Yes, this probably means duplicating the code from the CUT and putting it in the test, which would seem like a violation of the DRY principal, but in this case it would be appropriate, because what we want is independent confirmation of the result.  The problem with this is, we'll be tempted (now and in the future) to cut and paste the code from the CUT into the test, which will result in the test passing even if the CUT is wrong!  Thus, this is probably not a good approach.

That leaves us with hand-crafting the output from the input (which is what you've done).  And, the only question then is, where do we put the result of those labors.  And, my answer is, put it as close to the input as practical.  For instance, rather than having the mock generate the input in-line, I would move that definition to be a class attribute or a module-global which the mock simply references.  And then I would put the output definition right next to it, so that anybody who wants to compare them can find them both easily (and, you can put a single block comment explaining what the two of them are and how they are used).

Since you mention it, it might be cool/handy to have `find_template()` return the RHS value.  Since the fixture doesn't actually return any other value (it just sets up and later removes the mock), you probably _could_ do that.  (I'm not sure -- it kind of depends on how the PyTest fixture value caching works -- but it seems plausible.)  The problem is, though, you have set up the mock with an embedded conditional whose value isn't available at test-setup, so you would probably have to split the fixture into multiple fixtures, one for each value that the mock should return.  However, this might actually be a good thing to do anyway.  (That is, the mock could validate its argument to ensure that it's the expected request and then return its value which is hardwired to correspond to that request, which would obviate the need for the `else` clause that I requested above.)

Reply:
> Although I hesitate to suggest it, my instinct is that the second one should be dynamically generated by the test from the first one. 

I did something like this for publish; but since I was generating expected Elasticsearch responses from the document map, it wasn't directly duplicating Pbench code and that theoretical temptation doesn't exist. While "hand crafting" could be said to present a similar temptation, that is, copying and pasting the actual output of the Pbench code, the risk would be perhaps a bit smaller since it's easier to visually validate that the format is correct.

> Since you mention it, it might be cool/handy to have find_template() return the RHS value. Since the fixture doesn't actually return any other value (it just sets up and later removes the mock), you probably could do that.

Yes, you definitely *can* do that, and we've got several fixtures that do.

Reply:
Interesting, yeah I will fix this refactoring in another PR where I am already doing some test structure refactor.